src/mappings/helpers/helpers.ts

```diff
@@ -916,6 +961,68 @@ export function updateAdvancedIndexerMetrics(indexer: Indexer): Indexer {
   return indexer as Indexer
 }
 
+export function calculateOwnStakeRatioForProvision(provision: Provision): BigDecimal {
+  let totalTokens = provision.tokensProvisioned.plus(provision.delegatedTokens)
+  return totalTokens == BigInt.fromI32(0)
+    ? BigDecimal.fromString('0')
+    : provision.tokensProvisioned.toBigDecimal().div(totalTokens.toBigDecimal())
+}
+
+export function calculateDelegatedStakeRatioForProvision(provision: Provision): BigDecimal {
+  let totalTokens = provision.tokensProvisioned.plus(provision.delegatedTokens)
+  return totalTokens == BigInt.fromI32(0)
+    ? BigDecimal.fromString('0')
+    : provision.delegatedTokens.toBigDecimal().div(totalTokens.toBigDecimal())
+}
+
+export function calculateIndexingRewardEffectiveCutForProvision(provision: Provision): BigDecimal {
+  let delegatorCut =
+    provision.indexingRewardsCut.toBigDecimal() /
+    BigDecimal.fromString('1000000')
+  return provision.delegatedStakeRatio == BigDecimal.fromString('0')
+    ? BigDecimal.fromString('0')
+    : BigDecimal.fromString('1') - delegatorCut / provision.delegatedStakeRatio
+}
+
+export function calculateQueryFeeEffectiveCutForProvision(provision: Provision): BigDecimal {
+  let delegatorCut =
+    provision.queryFeeCut.toBigDecimal() / BigDecimal.fromString('1000000')
+  return provision.delegatedStakeRatio == BigDecimal.fromString('0')
+    ? BigDecimal.fromString('0')
+    : BigDecimal.fromString('1') - delegatorCut / provision.delegatedStakeRatio
+}
+
+export function calculateIndexerRewardOwnGenerationRatioForProvision(provision: Provision): BigDecimal {
+  let rewardCut =
+    provision.indexingRewardsCut.toBigDecimal() / BigDecimal.fromString('1000000')
+  return provision.ownStakeRatio == BigDecimal.fromString('0')
+    ? BigDecimal.fromString('0')
+    : rewardCut / provision.ownStakeRatio
+}
+
+export function calculateOverdelegationDilutionForProvision(provision: Provision): BigDecimal {
+  let provisionedTokensBD = provision.tokensProvisioned.toBigDecimal()
+  let delegatedTokensBD = provision.delegatedTokens.toBigDecimal()
+  let graphNetwork = GraphNetwork.load('1')!
+  let delegationRatioBD = BigInt.fromI32(graphNetwork.delegationRatio).toBigDecimal()
+  let maxDelegatedStake = provisionedTokensBD * delegationRatioBD
```

Comment:
I think this calculation is incorrect. The max useful delegated stake would be the provisioned tokens times the delegation ratio of the target data service. 

The delegation ratio is stored in the data service so we would need to query it from there, which probably means adding a `delegationRatio` property to DataService entity? When the delegation ratio is set an event is emitted on the data service contract: `event DelegationRatioSet(uint32 ratio)`.

Worth noting as well that delegation ratio is NOT MANDATORY, meaning a data service might consider all delegation as useful stake and not even have a delegation ratio. So we should probably initialize the delegationRatio as MAX uint256?

Let me know if you want to discuss this.

Reply:
Yeah, seems this one also slipped through the cracks, and the clarification makes sense, but the metric itself lacks any point if the data service allows all stake to be used, since that means there's no dilution (i.e. there can never be any extra delegation taking rewards/fees that they aren't helping generate, since useful delegation isn't capped on that case)